core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1043,6 +1050,43 @@ class KafkaApis(val requestChannel: RequestChannel,
     info("Shutdown complete.")
   }
 
+  def handleCreateTopicsRequest(request: RequestChannel.Request) {
+    val createTopicsRequest = request.body.asInstanceOf[CreateTopicsRequest]
+
+    val (validTopics, duplicateTopics) = createTopicsRequest.topics.asScala.partition { case (topic, _) =>
+      !createTopicsRequest.duplicateTopics.contains(topic)
+    }
+
+    def sendResponseCallback(results: Map[String, Errors]): Unit = {
+      val respHeader = new ResponseHeader(request.header.correlationId)
+      if(duplicateTopics.nonEmpty)
+        warn(s"Create topics request from client ${request.header.clientId} contains multiple entries for the following topics: ${duplicateTopics.keySet.mkString(",")}")
+      val completeResults = results ++ duplicateTopics.keySet.map((_, Errors.INVALID_REQUEST)).toMap
+      val responseBody = new CreateTopicsResponse(completeResults.asJava)
+      trace(s"Sending create topics response $responseBody for correlation id ${request.header.correlationId} to client ${request.header.clientId}.")
+      requestChannel.sendResponse(new RequestChannel.Response(request, new ResponseSend(request.connectionId, respHeader, responseBody)))
+    }
+
+    if(!controller.isActive()) {
+      val results = createTopicsRequest.topics.asScala.map { case (topic, _) =>
+        (topic, Errors.NOT_CONTROLLER)
+      }
+      sendResponseCallback(results)
```

Comment:
I need to compute validTopics/duplicateTopics before calling adminManager.createTopics so I don't create those topics, and then in the callback return a single instance with the error. 

The reason I don't use validTopics in the first 2 cases, when the controller is wrong or the user is not authorized is because I thought those exceptions take precedence over the duplicate topics exception.